optuna/_optimize.py

```diff
@@ -73,51 +72,51 @@ def _optimize(
             if show_progress_bar:
                 warnings.warn("Progress bar only supports serial execution (`n_jobs=1`).")
 
-            time_start = datetime.datetime.now()
-
-            def _should_stop() -> bool:
-                if study._stop_flag:
-                    return True
-
-                if timeout is not None:
-                    # This is needed for mypy.
-                    t: float = timeout
-                    return (datetime.datetime.now() - time_start).total_seconds() > t
+            if n_jobs == -1:
+                n_jobs = multiprocessing.cpu_count()
 
-                return False
-
-            if n_trials is not None:
-                _iter = iter(range(n_trials))
-            else:
-                _iter = iter(_should_stop, True)
-
-            with Parallel(n_jobs=n_jobs, prefer="threads") as parallel:
-                if not isinstance(
-                    parallel._backend, joblib.parallel.ThreadingBackend
-                ) and isinstance(study._storage, storages.InMemoryStorage):
-                    warnings.warn(
-                        "The default storage cannot be shared by multiple processes. "
-                        "Please use an RDB (RDBStorage) when you use joblib for "
-                        "multi-processing. The usage of RDBStorage can be found in "
-                        "https://optuna.readthedocs.io/en/stable/tutorial/rdb.html.",
-                        UserWarning,
-                    )
-
-                parallel(
-                    delayed(_optimize_sequential)(
-                        study,
-                        func,
-                        1,
-                        timeout,
-                        catch,
-                        callbacks,
-                        gc_after_trial,
-                        reseed_sampler_rng=True,
-                        time_start=time_start,
-                        progress_bar=None,
-                    )
-                    for _ in _iter
-                )
+            time_start = datetime.datetime.now()
+            n_submitted_trials = 0
+            futures: List[Future] = []
+
+            with ThreadPoolExecutor(max_workers=n_jobs) as executor:
+                while True:
+                    if study._stop_flag:
+                        break
+
+                    if (
+                        timeout is not None
+                        and (datetime.datetime.now() - time_start).total_seconds() > timeout
+                    ):
+                        break
+
+                    if n_trials is not None and n_submitted_trials >= n_trials:
+                        break
+
+                    for f in futures:
+                        if f.done():
+                            f.result()
+                            futures.remove(f)
+
+                    if len(futures) >= n_jobs:
+                        time.sleep(1)
```

Comment:
I think we can avoid having to sleep for a fixed number of seconds by introducing a block using [`wait`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait). Something like this.

```python
futures: Set[Future] = set()

with ThreadPoolExecutor(max_workers=n_jobs) as executor:
    for n_submitted_trials in itertools.count():
        if study._stop_flag:
            break

        if (
            timeout is not None
            and (datetime.datetime.now() - time_start).total_seconds() > timeout
        ):
            break

        if n_trials is not None and n_submitted_trials >= n_trials:
            break

        if len(futures) >= n_jobs:
            _, futures = wait(futures, return_when=FIRST_COMPLETED)

        futures.add(
            executor.submit(
                _optimize_sequential,
                study,
                func,
                1,
                timeout,
                catch,
                callbacks,
                gc_after_trial,
                True,
                time_start,
                None,
            )
        )
```

Reply:
Thanks for your insightful suggestion. I applied.